Add support for the anyio library

jakkdl opened the pull request:
Fixes #61
Have fun with this tiny commit :upside_down_face: 

Kind of a chore, I maybe should have gone at some parts in different ways, but the library should now have full support for anyio - and all eval files are tested with "trio" replaced with "anyio", looking to see that all the errors are still good, and that there's no instances of "trio" in the output.

If you trigger an error without importing either trio or anyio, the library will default to saying "trio" in error messages. If you trigger an error having imported both, it will say stuff like
> sync call ... in async function, use `[trio|anyio].open_file(...)`.

Otherwise it's not too much that's actually changed, just lots of small replacements in a billion places.

Zac-HD commented on flake8_trio/visitors/visitor103_104.py:
```diff
@@ -70,10 +71,11 @@ def visit_ExceptHandler(self, node: ast.ExceptHandler):
             extra_message = ""
-            if marker.name != "trio.Cancelled":
+            if not re.fullmatch(r"(trio|anyio)\.Cancelled", marker.name):
+                library = self.library if self.library else "trio"
                 extra_message = (
-                    " Consider adding an `except trio.Cancelled: raise` before this"
-                    " exception handler."
+                    f" Consider adding an `except {library}.Cancelled: raise`"
+                    " before this exception handler."
```
See `Cancelled` comment above.

Zac-HD commented on flake8_trio/visitors/visitors.py:
```diff
@@ -127,35 +128,38 @@ def visit_Call(self, node: ast.Call):
-                (node.func.value.id == "trio" and node.func.attr in trio_async_funcs)
+                (
+                    node.func.value.id in ("trio", "anyio")
+                    and node.func.attr in trio_async_funcs
```
Some of `trio_async_funcs` don't exist in `anyio`, and vice-versa; we should probably make a second list (well, tuple) for `anyio` and check them separately.

jakkdl replied:
oh hum, it also doesn't have `.open_nursery()`

Zac-HD commented on flake8_trio/visitors/visitors.py:
```diff
@@ -386,12 +394,16 @@ class Visitor117(Flake8TrioVisitor):
     def visit_Attribute(self, node: ast.Attribute):
-        if (n := ast.unparse(node)) in ("trio.MultiError", "trio.NonBaseMultiError"):
-            self.error(node, n)
+        if (
+            isinstance(node.value, ast.Name)
+            and node.value.id in ("trio", "anyio")
+            and node.attr in DEPRECATED_ERRORS
+        ):
+            self.error(node, ast.unparse(node))
```
`anyio` doesn't have the old `MultiError`, just `ExceptionGroup` 🙂

Zac-HD commented on flake8_trio/visitors/flake8triovisitor.py:
```diff
@@ -117,3 +118,12 @@ def save_state(self, node: ast.AST, *attrs: str, copy: bool = False):
         for b in body:
             yield from ast.walk(b)
+
+    def visit_Import(self, node: ast.Import):
+        for alias in node.names:
+            name = alias.name
+            if name in ("trio", "anyio") and alias.asname is None:
+                if self.library is not None and name != self.library:
+                    self.library = f"[{self.library}|{name}]"
+                else:
+                    self.library = alias.name
```
This is great, but I think we should also have a global `--anyio` flag to set the default library; otherwise `anyio` projects will have `trio` recommended to them in any file that doesn't happen to import `anyio`!

Zac-HD commented on flake8_trio/visitors/helpers.py:
```diff
@@ -103,10 +103,10 @@ def has_exception(node: ast.expr | None) -> str:
             and isinstance(node.value, ast.Name)
-            and node.value.id == "trio"
+            and node.value.id in ("trio", "anyio")
             and node.attr == "Cancelled"
         ):
-            return "trio.Cancelled"
+            return f"{node.value.id}.Cancelled"
```
Unfortunately `anyio` has a *slightly* different API here; see [`anyio.get_cancelled_exc_class()`](https://anyio.readthedocs.io/en/stable/api.html#anyio.get_cancelled_exc_class).  That's important because you can have both `trio` and `asyncio` in the same program (ugh, but sometimes the least-bad option!).

We probably also want to lint for "no assigning that result because it breaks our linter, as well as multi-backend programs".

jakkdl replied:
I assume you mean `anyio` and not `asyncio`? Maybe not the best name they could've chosen :sweat_smile: 

I'll probably push this to a separate issue+PR.

jakkdl replied:
:scream:

jakkdl replied:
Tracked in #122

Zac-HD reviewed (commented):
This is *shockingly* easy for such a conversion - huge kudos to you for an implementation that made this feasible, and @agronholm for a super-clean drop in library 🙌

A few things before this is ready to ship are listed in comments below; happy to do those here or in a follow-up PR as you prefer.

🤔 I should probably also run this past @agronholm too, in case I missed something...

jakkdl commented:
Fixed all review comments (or disabled functionality and moved out to separate issues). Also did some refactoring to make stuff cleaner, added `# NOTRIO` and `# ANYIO_NO_ERROR` (those magic markers should perhaps have some checks that they're not misspelled or something, or use some other way of being more robust)

Zac-HD commented on tests/eval_files/trio105_anyio.py:
```diff
@@ -0,0 +1,13 @@
+
+
+async def foo():
+    nursery = anyio.open_nursery()
```
Probably worth a comment that all this is fine because anyio calls the equivalent construct a "taskgroup", following asyncio?

agronholm replied:
Following curio, actually. Asyncio got task groups much later than either curio or AnyIO.

Zac-HD commented on tests/eval_files/trio102_anyio.py:
```diff
@@ -0,0 +1,28 @@
+        ...
+    except ValueError:
+        await foo()  # safe
+    except anyio.Cancelled:
```
I presume this will be updated in a future PR 👍

agronholm replied:
Yeah, there is no `anyio.Cancelled`.

agronholm replied:
The AnyIO way to do this is `except anyio.get_cancelled_exc_class():`.

Zac-HD replied:
Yep, discussed in https://github.com/Zac-HD/flake8-trio/pull/120#discussion_r1095001688 and tracked in #122 🙂

jakkdl replied:
Yeah that check is to make sure it doesn't raise an error - regardless of if it exists as a library function or not. Can add a comment with #122
